### ISSUE TITLE
Control and Alt keys not working for Vimium Chrome on MacOS

### ISSUE THREAD

**Issue by datavistics** [philc/vimium#3620]
**Describe the bug**
Control and Alt keys not working for Vimium Chrome on MacOS. 

**To Reproduce**
c-f is supposed to go down a page.
a-f is supposed to open multiple links in new tabs
When I try command + f, I get the browser search. Im not sure how to use alt.

**Browser and Vimium version**
Vimium: Version 1.66
Brave 1.8.96 (though same with chrome)

**Comment by gdh1995** [comment]
For linkhints mode, you need to press and hold on <kbd>Ctrl</kbd>, and then press a letter key to activate it.

**Comment by isaiahtaylorhh** [comment]
I wasn't able to get this to work, either alt-f, command-f, or holding down ctrl.

**Comment by gdh1995** [comment]
press a single f to open linkhints mode, and then press down and hold on ctrl and press some hint characters, to open a link in the new tab.

**Comment by datavistics** [comment]
@gdh1995 , I think you are answering a different question. I know how to follow a link. 

Can you tell me on a mac using Vimium how to scroll down a page? According to the documentation its Cmd-f. But Cmd-f opens the browser find.

**Comment by gdh1995** [comment]
Why do you think it's a meta-f? The default key of scrollding down is a single J

**Comment by gdh1995** [comment]
By the way, c-f means to press control and f, And m-f means meta(command) and f, while a-f is only for windows and linux - There is no alt key on a macOS keyboard layout.

**Comment by datavistics** [comment]
Let me start over, I think we got lost somewhere.


According to this [site](https://vimium.github.io/): 
c-f | Scroll a full page down
c-b | Scroll a full page up
I know about `j` and `d` to **scroll down** and **scroll a half page down** respectively. 
Id like to know how to scroll a full page down. 
Its possible that these are dates as they are from the github.io site. 

Id also like to know how to 
`<a-p>   pin/unpin current tab`
`<a-f>   open multiple links in a new tab`
Which are taken from the [README](https://github.com/philc/vimium). 

Im on a Mac, so I dont have an Alt key. Ive tried Shift, Option, Control, and Command with no luck.

**Comment by gdh1995** [comment]
Um the page is too old - according to https://github.com/philc/vimium/blob/master/CHANGELOG.md, `<c-f>` has been removed from the default binding list since v1.26 (2011-02-17).

You may map other keys to `pin/unpin current tab`. For example, `<m-p>` means Meta + P, and `P` means Shift + P

**Comment by datavistics** [comment]
Ok, good to know about the c-f. 

I can remap easy enough, and that is a good solution for me. 

Id like to know what the intended behavior is for Macs though. There should be a disclaimer saying windows only, or maybe there is a bug in some of the modifier keys for macs.

**Comment by gdh1995** [comment]
There're only 3 default key bindings using Alt: a-f LinkHints.activateModeWithQueue, a-p togglePinTab, a-m toggleMuteTab , and usually users will remap keys when they want to use some other commands.

Vimium should work well on Mac, because its main developers seem to have macOS for daily use. If a composed key can not be grabbed by Vimium, the most common reason is that it's a high-priority shortcut of Chrome and can not be overridden by JavaScript.

**Comment by gdh1995** [comment]
There're only 3 default key bindings using Alt: a-f LinkHints.activateModeWithQueue, a-p togglePinTab, a-m toggleMuteTab , and usually users will remap keys when they want to use some other commands.

Vimium should work well on Mac, because its main developers seem to have macOS for daily use. If a composed key can not be grabbed by Vimium, the most common reason is that it's a high-priority shortcut of Chrome and can not be overridden by JavaScript.

**Comment by isaiahtaylorhh** [comment]
Just conforming @datavistics: I just got this extension yesterday and have had no luck with a-f. I'm on Mac. Similarly I have tried command, option, control, shift, fn, and combinations of the above.

**Comment by templateK** [comment]
@datavistics @isaiahtaylorhh  Have you enabled the ` Ignore keyboard layout` in `Advanced Options`?

https://github.com/philc/vimium/issues/3197#issuecomment-614829140

> Vimium is unable to detect mappings using Option (Alt) key as modifier on macOS unless "Ignore keyboard layout" setting is checked. Otherwise it uses `KeyboardEvent.key` to identify the key press which means e.g. `<a-p>` is identified as letter π on macOS since Option key is [used to type additional characters](https://en.wikipedia.org/wiki/Option_key#Use_of_the_Option_key). See relevant code:
> 
> https://github.com/philc/vimium/blob/ac5f9d3407bcf07c50a0119aede916fdd561f224/lib/keyboard_utils.js#L22-L23

This should be documented on somewhere if it is not.  
I think the best place is the note section of the `Ignore keyboard layout` option.

**Comment by templateK** [comment]
FWIW, I noticed that option key binding works well without enabling the `Ignore keyboard layout` option on macOS firefox 
which contradicts https://github.com/philc/vimium/issues/3197#issuecomment-614829140

**Comment by datavistics** [comment]
@templateK 
That worked! This should be broadcasted more visibly. I looked for quite a while and didnt find anything. 
Can we put this in the readme for mac users?

**Comment by politicus** [comment]
@datavistics +100 

Vimium is a productivity extension which means sparing time. It is a fantastic extension and the creator(s?), maintainer(s?) is (are?) doing a fantastic job! But the documentation could be easily and greatly improved especially to cover these very specific kind of cases. 

This is not the first time I spend a significant amount of time to find the solution to a "bug" like this one: changed the mapping of one key of my Vimium settings for the 50th time and nothing happened. Tried new mappings with other Alt-key combinations. Same result. Went to Vimium wiki. Found nothing. Searched this repository. Finally found the solution. 

If I were proficient enough in git, I would definitely create some some PRs.

**Comment by ranelpadon** [comment]
I've experienced the same issue in my Mac-FF because I have symbols/punctuation layer (AltGr) set in my Colemak layout as well as in Karabiner Elements. For example, the `search` (`/`) char is triggered/outputted by `Alt-l` in my customization which will not work because I couldn't enable this setting since I use Colemak layout: `Ignore keyboard layout option`. 

I think `Alt` should be supported regardless of keyboard layout/customization since mine is working as intended in all apps including `vim`. It doesn't work only when using `Vimium`.

**Comment by resting** [comment]
Was looking for this as well. This binding worked for me on mac.
```
map <c-f> scrollFullPageDown
map <c-b> scrollFullPageUp
```

**Comment by Omixxx** [comment]
For those who struggle as me, i was able to use `  Alt + "key"` succesfully by disabling the previous mapped shortcuts and map using the following notation


````
unmap J
unmap K
unmap gt
unmap gT

map <a-h> previousTab
map <a-l> nextTab 
````

A Big thanks to you devs! It's an awesome tool  🧑‍💻

**Comment by philc** [comment]
Thanks for all of the discussion everyone. This is a tricky issue and I'll look into a fix. I'll close this in favor of #3197, which summarizes the issue.

**Comment by torma616** [comment]
I know this is closed and I left the same comment on #3197, but for anybody else who finds this one first via google like I did:

In summary - without having to `Ignore Keyboard Layout` just `map <a-{the symbol produced by option + key}> command` in your Vimium options.
In Vimium options, simply map the output of Option + Letter to <a-{SYMBOL}>

In other words:
Vimium Default:
`<a-f> LinkHints.activateModeWithQueue`

On macOS:
`Alt(Option)+f` produces `ƒ`

In Vimium Custom Key Mappings: 
`map <a-ƒ> LinkHints.activateModeWithQueue`

https://github.com/philc/vimium/issues/3197#issuecomment-2701499493
